src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+#include "utils/mutex.h"
+#include "v8_engine/script.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/hash/hash.h>
+
+#include <exception>
+#include <iostream>
+#include <variant>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+template<typename Executor>
+class script_dispatcher {
+    constexpr static size_t _mutex_count{50};
+
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _mutexes(_mutex_count)
+      , _executor(executor) {}
+
+    ss::future<>
+    update(model::topic_namespace& topic, script_compile_params script_params) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.insert_or_assign(topic, std::move(script_params));
+    }
+
+    ss::future<> remove(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.erase(topic);
+    }
+
+    // TODO: need to think about linearizability
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
```

Comment:
Do I need gate there?

Reply:
a gate is for controlled shutdown. Usually classes which dispatch async work use a gate to ensure no async work continues after some point (usually after a method named `stop()` resolves). I would say this needs a gate